opentelemetry/src/context.rs

```diff
@@ -381,10 +388,107 @@ impl Hasher for IdHasher {
     }
 }
 
+/// A stack for keeping track of the [`Context`] instances that have been attached
+/// to a thread.
+///
+/// The stack allows for popping of contexts by position, which is used to do out
+/// of order dropping of [`ContextGuard`] instances. Only when the top of the
+/// stack is popped, the topmost [`Context`] is actually restored.
+///
+/// The stack relies on the fact that it is thread local and that the
+/// [`ContextGuard`] instances that are constructed using it can't be shared with
+/// other threads.
+struct ContextStack {
+    /// This is the current [`Context`] that is active on this thread, and the top
+    /// of the [`ContextStack`]. It is always present, and if the `stack` is empty
+    /// it's an empty [`Context`].
+    ///
+    /// Having this here allows for fast access to the current [`Context`].
+    current_cx: Context,
+    /// A `stack` of the other contexts that have been attached to the thread.
+    stack: Vec<Option<Context>>,
+    /// Ensure this type is !Send as it relies on thread locals
+    _marker: PhantomData<*const ()>,
+}
+
+impl ContextStack {
+    const BASE_POS: u16 = 0;
+    const MAX_POS: u16 = u16::MAX;
+    const INITIAL_CAPACITY: usize = 8;
+
+    #[inline(always)]
+    fn push(&mut self, cx: Context) -> u16 {
+        // The next id is the length of the `stack`, plus one since we have the
+        // top of the [`ContextStack`] as the `current_cx`.
+        let next_id = self.stack.len() + 1;
+        if next_id < ContextStack::MAX_POS.into() {
+            let current_cx = std::mem::replace(&mut self.current_cx, cx);
+            self.stack.push(Some(current_cx));
+            next_id as u16
+        } else {
+            // This is an overflow, log it and ignore it.
+            otel_warn!(name: "ContextStack.push", message = "Context stack overflow, context not pushed.");
```

Comment:
```suggestion
            otel_warn!(name: "Context.AttachFailed", message = "Too many contexts. Max limit is {}", ContextStack::MAX_POS);
```


^ Something like this. ContextStack is an internal implementation which can be changed anytime, so for user-facing error message, its best to avoid mentioning it.

Reply:
Also good to mention that the Context::Current will be the context that was attempted to be pushed, so it must be dropped to return to previous Context. (Is my understanding correct?)

Reply:
When this happens, the `Context` that we're trying to `attach` will not be pushed, so `Context::current()` will be the same `Context` as before this `attach` call. I'm not sure how to convey this in a concise manner in a log line.

Reply:
Got it. Then we can state "Context::current() remains unchanged as this attach failed. Dropping the returned ContextGuard will also have no impact on Context::current()"